### PR TITLE
Authentication

### PR DESCRIPTION
**Pull request by CharlieMantium**

### CODE REVIEW

File: src/components/ActivitiesListItem/ActivitiesListItem.js [hunk 1]
```diff
@@ -74,7 +74,7 @@ ActivitiesListItem.defaultProps = {
   filter: '',
 };
 
-const mapStateToProps = ({ filter }) => ({
+const mapStateToProps = ({ activities: { filter } }) => ({
```
Comment (Lukasz-Semik) [inline src/components/ActivitiesListItem/ActivitiesListItem.js]: this name is very problematic. When dev hears filter, he is thinking about filter method. I think, that it would be a nice improvement to change it somehow (maybe `filterKey` or something like that?)

File: src/components/Header/Header.js [hunk 2]
```diff
@@ -79,18 +81,21 @@ const Header = ({ onSetFilter }) => {
         <NavLink to={CREATE_ACTIVITY_ROUTE} data-test="react-navlink">
           <AddIcon title="Add New Activity" />
         </NavLink>
+        <Button onClick={onStartLogout}>Logout</Button>
```
Comment (Lukasz-Semik) [inline src/components/Header/Header.js]: what about translation?

File: src/components/Header/Header.js [hunk 3]
```diff
@@ -79,18 +81,21 @@ const Header = ({ onSetFilter }) => {
...
 
 Header.propTypes = {
   onSetFilter: PropTypes.func.isRequired,
+  onStartLogout: PropTypes.func.isRequired,
 };
 
-const mapDispatchToProps = {
-  onSetFilter: setFilter,
-};
+const mapDispatchToProps = dispatch => ({
```
Comment (Lukasz-Semik) [inline src/components/Header/Header.js]: what is the reason of this change to function? It will work `normally` if object will be used 🤔

File: src/components/Header/Header.test.js [hunk 4]
```diff
@@ -6,7 +6,10 @@ import { HeaderUnwrapped } from './Header';
 describe('Header', () => {
   it('should render Header correctly', () => {
     const onSetFilterSpy = jest.fn();
-    const wrapper = shallow(<HeaderUnwrapped onSetFilter={onSetFilterSpy} />);
+    const onStartLogout = jest.fn();
+    const wrapper = shallow(
+      <HeaderUnwrapped onSetFilter={onSetFilterSpy} onStartLogout={onStartLogout} />,
```
Comment (Lukasz-Semik) [inline src/components/Header/Header.test.js]: why there is no assertion added to verify if this function has been invoked properly when something is clicked?

File: src/components/LoginPage/LoginPage.js [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+    <Button onClick={onStartLogin}>Login</Button>
+  </LoginWrapper>
+);
+
+LoginPage.propTypes = {
+  onStartLogin: PropTypes.func.isRequired,
+};
+
+const mapDispatchtoProps = dispatch => ({
+  onStartLogin: () => dispatch(startLogin()),
```
Comment (Lukasz-Semik) [inline src/components/LoginPage/LoginPage.js]: the same as before. By the way... you can use `useDispatch` ;) espacially in such a small function. 
in `mponents/Header/Header.js` as well

File: src/index.js [hunk 6]
```diff
@@ -2,10 +2,34 @@ import React from 'react';
...
+
+firebase.auth().onAuthStateChanged(user => {
+  if (user) {
+    store.dispatch(login(user.uid));
+    renderApp();
+    if (history.location.pathname === LOGIN_PAGE_ROUTE) {
+      history.push(ACTIVITY_PLAN_ROUTE);
+    }
+  } else {
+    store.dispatch(logout());
```
Comment (Lukasz-Semik) [inline src/index.js]: why there is an logout? I mean, if there is no user, there is no one to be logout 🤔

File: src/router/PrivateRoute.js [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+import { connect } from 'react-redux';
+import { Route, Redirect } from 'react-router-dom';
+
+import Header from '../components/Header';
+import { LOGIN_PAGE_ROUTE } from '../constants/routes';
+
+export const PrivateRoute = ({ isAuthenticated, component: Component, ...rest }) => (
+  <Route
+    {...rest}
+    component={props => {
```
Comment (Lukasz-Semik) [inline src/router/PrivateRoute.js]: you are passing a function, so you should use `render` prop. `component` and `render` has differences. 
also:
`props => isAuthenticated ? <><Header />.... </> : <Redirect>` 
I am not sure about this approach at all, but at this moment let's stay as is if works correctly

File: src/router/PrivateRoute.js [hunk 8]
```diff
@@ -0,0 +1,35 @@
...
+  />
+);
+
+PrivateRoute.propTypes = {
+  isAuthenticated: PropTypes.bool.isRequired,
+  component: PropTypes.elementType.isRequired,
+};
+
+const mapStateToProps = state => ({
+  isAuthenticated: !!state.auth.uid,
```
Comment (Lukasz-Semik) [inline src/router/PrivateRoute.js]: to be more modern, let's use `Boolean` constructor

File: src/router/PublicRoute.js [hunk 9]
```diff
@@ -0,0 +1,29 @@
+import React from 'react';
+import PropTypes from 'prop-types';
+import { connect } from 'react-redux';
+import { Route, Redirect } from 'react-router-dom';
+
+import { ACTIVITY_PLAN_ROUTE } from '../constants/routes';
+
+export const PublicRoute = ({ isAuthenticated, component: Component, ...rest }) => (
+  <Route
```
Comment (Lukasz-Semik) [inline src/router/PublicRoute.js]: similar to previous one

File: src/store/activities/reducer.test.js [hunk 10]
```diff
@@ -80,12 +77,12 @@ describe('activities reducer', () => {
     expect(state).toEqual(state);
   });
 
-  it('should set expenses', () => {
```
Comment (Lukasz-Semik) [inline src/store/activities/reducer.test.js]: tooooooo much making the copies instead of writing it by your own ! 😠

File: src/store/auth/actions.js [hunk 11]
```diff
@@ -0,0 +1,24 @@
...
+  return () => {
+    return firebase.auth().signInWithPopup(googleAuthProvider);
+  };
+};
+
+export const logout = () => ({
+  type: LOGOUT,
+});
+
+export const startLogout = () => {
```
Comment (Lukasz-Semik) [inline src/store/auth/actions.js]: make it shorter and prettier, this is an arrow function ;)

File: src/store/auth/reducer.js [hunk 12]
```diff
@@ -0,0 +1,14 @@
+import { LOGIN, LOGOUT } from './actionTypes';
+
+export default (state = {}, action) => {
+  switch (action.type) {
+    case LOGIN:
+      return {
```
Comment (Lukasz-Semik) [inline src/store/auth/reducer.js]: you are  breaking a convention of not mutating the state here. There is no taking care about other (at this moment not existing fields)

File: yarn.lock [hunk 13]
```diff
@@ -5606,6 +5606,18 @@ hash.js@^1.0.0, hash.js@^1.0.3:
     inherits "^2.0.3"
     minimalistic-assert "^1.0.1"
 
+history@^4.10.1:
```
Comment (Lukasz-Semik) [inline yarn.lock]: try to use `BrowserRouter` from `react-router-dom` instead of playing with this package.

File: src/components/Header/Header.test.js (hunk 4 above)
Comment (CharlieMantium) [reply src/components/Header/Header.test.js]: this is one of those tricky-son-of-a-bitch test that I can't handle. I will get back to it in future PRs related specifically to tests.

File: src/index.js (hunk 6 above)
Comment (CharlieMantium) [reply src/index.js]: There is no login in firebase (as you said no one is loged in) but this logout() action removes login from store and app is rendered properly.

File: src/store/activities/reducer.test.js (hunk 10 above)
Comment (CharlieMantium) [reply src/store/activities/reducer.test.js]: :( just to explain myself, I REALLY don't copy stuff from somewhere else, it would be a waste of my and your time. Sometimes when I'm stuck I search for answers in this React Udemy course and as it is concentrated on making this EXPENSES app and I really try to understand what they are doing and I focus extremely I catch myself doing stupid mistakes like this. Implementing things I just understood often I just rush with the flow and it ends up like this :/

File: yarn.lock (hunk 13 above)
Comment (CharlieMantium) [reply yarn.lock]: yes, I had a <BrowserRouter> there but when I needed to history.push in file index.js I started to have some problems and I found this workaround to change it to <Router> and install this 'history' package. I remember that I spend some frustrating time on this, however if you can give me a quick tip how to get back to <BrowserRouter> and avoid issues with history.push as I mentioned - I will be more than grateful.

Comment (Lukasz-Semik) [reply yarn.lock]: hm. Ok, this is not wrong solution you created, without some kind of pairing I am not able to give you a hint 🤔 Let's stay as is for now

File: src/components/Header/Header.js [hunk 14]
```diff
@@ -58,7 +58,11 @@ const AddIcon = styled(DiffAdded)`
   }
 `;
 
-const Header = ({ onSetFilter, onStartLogout }) => {
+const Header = () => {
+  const dispatch = useDispatch();
+  const onSetFilter = textValue => dispatch(setFilter(textValue));
+  const onStartLogout = dispatch(startLogout);
```
Comment (Lukasz-Semik) [inline src/components/Header/Header.js]: more common solution is to pass function to be dispatched, not the result of dispatch (in similar way as `onSetFilter`. Is there some special reason for this solution?

File: src/components/LoginPage/LoginPage.js [hunk 15]
```diff
@@ -1,29 +1,19 @@
...
-const LoginPage = ({ onStartLogin }) => (
-  <LoginWrapper>
-    <Button onClick={onStartLogin}>Login</Button>
-  </LoginWrapper>
-);
-
-LoginPage.propTypes = {
-  onStartLogin: PropTypes.func.isRequired,
+const LoginPage = () => {
+  const onStartLogin = useDispatch()(startLogin);
```
Comment (Lukasz-Semik) [inline src/components/LoginPage/LoginPage.js]: let's do as facebook wants to 
```
 const dispatch = useDispatch();
const onStartLogin = () => dispatch(startLogin())
```

File: src/components/LoginPage/LoginPage.js [hunk 16]
```diff
@@ -1,29 +1,19 @@
...
+  return (
+    <LoginWrapper>
+      <Button onClick={onStartLogin}>Login</Button>
```
Comment (Lukasz-Semik) [inline src/components/LoginPage/LoginPage.js]: what about translation?

File: src/router/PublicRoute.js [hunk 17]
```diff
@@ -8,7 +8,7 @@ import { ACTIVITY_PLAN_ROUTE } from '../constants/routes';
 export const PublicRoute = ({ isAuthenticated, component: Component, ...rest }) => (
   <Route
     {...rest}
-    component={props => {
+    render={props => {
```
Comment (Lukasz-Semik) [inline src/router/PublicRoute.js]: why not tenary?

Comment (CharlieMantium) [reply src/router/PublicRoute.js]: I used 'if' intentionally. There is similar component in PrivateRoute and there I have this annoying eslint-prettier issue with indentation in ternary operator broken into multiple lines. Last time you said to ignore it and use 'if' instead of fighting with configuration. In this file ternary operator is working correctly but i changed both to 'if' for consistency.
